vpn/speaker.go

```diff
@@ -0,0 +1,368 @@
+// struct and sends them over the requests chan.  The manager/tunnel role type must read from this chan and handle
+// the requests.  If they are RPC types, it should call sendReply() on the request with the reply message.
+type speaker[S rpcMessage, R receivableRPCMessage[RR], RR any] struct {
+	serdes    *serdes[S, R, RR]
+	requests  chan *request[S, R]
+	logger    slog.Logger
+	nextMsgID uint64
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	sendCh       chan<- S
+	recvCh       <-chan R
+	recvLoopDone chan struct{}
+
+	mu            sync.Mutex
+	responseChans map[uint64]chan R
+}
+
+// newSpeaker creates a new protocol speaker.
+func newSpeaker[S rpcMessage, R receivableRPCMessage[RR], RR any](
+	ctx context.Context, logger slog.Logger, conn io.ReadWriteCloser,
+	me, them SpeakerRole,
+) (
+	*speaker[S, R, RR], error,
+) {
+	ctx, cancel := context.WithCancel(ctx)
+	if err := handshake(ctx, conn, logger, me, them); err != nil {
+		cancel()
+		return nil, xerrors.Errorf("handshake failed: %w", err)
+	}
+	sendCh := make(chan S)
+	recvCh := make(chan R)
+	s := &speaker[S, R, RR]{
+		serdes:        newSerdes[S, R, RR](ctx, logger, conn, sendCh, recvCh),
+		logger:        logger,
+		requests:      make(chan *request[S, R]),
+		responseChans: make(map[uint64]chan R),
+		nextMsgID:     1,
+		ctx:           ctx,
+		cancel:        cancel,
+		sendCh:        sendCh,
+		recvCh:        recvCh,
+		recvLoopDone:  make(chan struct{}),
+	}
+	return s, nil
+}
+
+// start starts the serialzation/deserialization.  It's important this happens
+// after any assignments of the speaker to its owning Tunnel or Manager, since
+// the mutex is copied and that is not threadsafe.
+// nolint: revive
+func (s *speaker[_, _, _]) start() {
+	s.serdes.start()
+	go s.recvFromSerdes()
+}
+
+func (s *speaker[S, R, _]) recvFromSerdes() {
+	defer close(s.recvLoopDone)
+	defer close(s.requests)
+	for {
+		select {
+		case <-s.ctx.Done():
+			s.logger.Debug(s.ctx, "recvFromSerdes context done while waiting for proto", slog.Error(s.ctx.Err()))
+			return
+		case msg, ok := <-s.recvCh:
+			if !ok {
+				s.logger.Debug(s.ctx, "recvCh is closed")
+				return
+			}
+			rpc := msg.GetRpc()
+			if rpc != nil && rpc.ResponseTo != 0 {
+				// this is a unary response
+				s.tryToDeliverResponse(msg)
+				continue
+			}
+			req := &request[S, R]{
+				ctx:     s.ctx,
+				msg:     msg,
+				replyCh: s.sendCh,
+			}
+			select {
+			case <-s.ctx.Done():
+				s.logger.Debug(s.ctx, "recvFromSerdes context done while waiting for request handler", slog.Error(s.ctx.Err()))
+				return
+			case s.requests <- req:
+			}
+		}
+	}
+}
+
+// nolint: revive
+func (s *speaker[_, _, _]) Close() error {
+	s.cancel()
+	err := s.serdes.Close()
+	return err
+}
+
+// unaryRPC sends a request/response style RPC over the protocol, waits for the response, then
+// returns the response
+func (s *speaker[S, R, _]) unaryRPC(ctx context.Context, req S) (resp R, err error) {
```

Comment:
I may be misunderstanding but it seems like this is missing a condition on `serdes` unexpectedly exiting? Let's say your `serdes.SendLoop` unexpectedly exits because your connection dies when you try to write something to the wire, how does closure of your serializer propagate to the speaker here? Presumable we could also select on `recvLoopDone` and return an error?

Reply:
Yeah, that's a good idea.

Reply:
Added a test case too.